src/EFCore.PG/Query/ExpressionTranslators/Internal/NpgsqlBinaryStringTranslator.cs

```diff
@@ -0,0 +1,153 @@
+// AND FITNESS FOR A PARTICULAR PURPOSE. THE SOFTWARE PROVIDED HEREUNDER IS
+// ON AN "AS IS" BASIS, AND THE NPGSQL DEVELOPMENT TEAM HAS NO OBLIGATIONS
+// TO PROVIDE MAINTENANCE, SUPPORT, UPDATES, ENHANCEMENTS, OR MODIFICATIONS.
+
+#endregion
+
+using System;
+using System.Linq;
+using System.Linq.Expressions;
+using System.Reflection;
+using System.Security.Cryptography;
+using JetBrains.Annotations;
+using Microsoft.EntityFrameworkCore;
+using Microsoft.EntityFrameworkCore.Query.Expressions;
+using Microsoft.EntityFrameworkCore.Query.ExpressionTranslators;
+
+namespace Npgsql.EntityFrameworkCore.PostgreSQL.Query.ExpressionTranslators.Internal
+{
+    /// <summary>
+    /// Provides translation services for PostgreSQL binary string functions and operators.
+    /// </summary>
+    /// <remarks>
+    /// See: https://www.postgresql.org/docs/current/static/functions-binarystring.html
+    /// </remarks>
+    public class NpgsqlBinaryStringTranslator : IMethodCallTranslator
+    {
+        #region MethodInfo
+
+        [NotNull] static readonly MethodInfo ComputeHash1 =
+            typeof(HashAlgorithm).GetRuntimeMethod(nameof(HashAlgorithm.ComputeHash), new[] { typeof(byte[]) });
+
+        [NotNull] [ItemNotNull] static readonly MethodInfo[] SupportedCreateMethods =
+        {
+            typeof(MD5).GetRuntimeMethod(nameof(MD5.Create), Type.EmptyTypes),
+            typeof(SHA256).GetRuntimeMethod(nameof(SHA256.Create), Type.EmptyTypes),
+            typeof(SHA384).GetRuntimeMethod(nameof(SHA384.Create), Type.EmptyTypes),
+            typeof(SHA512).GetRuntimeMethod(nameof(SHA512.Create), Type.EmptyTypes),
+        };
+
+        #endregion
+
+        /// <summary>
+        /// The backend version to target.
+        /// </summary>
+        [CanBeNull] readonly Version _postgresVersion;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="NpgsqlBinaryStringTranslator"/> class.
+        /// </summary>
+        /// <param name="postgresVersion">The backend version to target.</param>
+        public NpgsqlBinaryStringTranslator([CanBeNull] Version postgresVersion)
+            => _postgresVersion = postgresVersion;
+
+        /// <inheritdoc />
+        [CanBeNull]
+        public Expression Translate(MethodCallExpression e)
+        {
+            // TODO: disable when version is unspecified until 11.0 is released.
+            if (_postgresVersion == null)
+                return null;
+
+            if (SupportedCreateMethods.Contains(e.Method))
+                return e;
+
+            return TranslateComputeHash1(e) ?? TranslateExtensions(e);
+        }
+
+        [CanBeNull]
+        Expression TranslateComputeHash1([NotNull] MethodCallExpression e)
+        {
+            if (e.Method != ComputeHash1)
+                return null;
+
+            switch (e.Object?.Type.Name)
+            {
+            case nameof(MD5) when VersionAtLeast(8, 1):
```

Comment:
So if I'm understanding this correctly, you're providing *two* ways to do MD5 - via the your special extension method and via the built-in MD5 class... Why not drop the former then?

Reply:
Then you need to reference the crypto API library even if you don't use it on the client side

Reply:
At least in the full framework (and likely for .NET Core), touching certain crypto APIs on a client with FIPS restrictions throws an exception.

I need to do some more testing to see if the evaluation filter above prevents that exception from triggering. If it does, we could remove all but the SHA224 extension (which I'm not seeing in the crypto API).

But if not these provider-specific functions would let users avoid the crypto exception altogether.

Reply:
> At least in the full framework (and likely for .NET Core), touching certain crypto APIs on a client with FIPS restrictions throws an exception.

But assuming the function actually executes server-side, why would you get any FIPS-related exception? You're not actually executing the function client-side...

But another thought is that if you're running with FIPS restrictions, maybe it's correct for calls to MD5 to fail with an exception. After all, the whole point of FIPS is to make sure you *don't* use stuff like MD5, right?

@YohDeadfall it's true that you need to reference the crypto API, but that seems totally OK to me... Whether you run it locally or server-side, you're using a crypto API and it doesn't seem unreasonable to reference that package...